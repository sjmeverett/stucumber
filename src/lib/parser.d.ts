
export interface Feature {
  name: Clause;
  scenarios: Scenario[];
  annotations: string[];
}

export interface Scenario {
  name: Clause;
  rules: Rule[];
  annotations: string[];
}

export interface Rule extends Clause {
  data: any;
}

export interface Clause {
  value: string;
  location: {
    line: number;
    column: number;
    offset: number;
  }
}

export function parse(source: string): Feature;
