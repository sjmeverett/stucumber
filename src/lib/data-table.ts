export default class DataTable {
  constructor(private data: string[][]) {}

  raw() {
    return this.data;
  }

  asObjects() {
    if (this.data.length < 1) {
      throw new Error('need at least 1 row');
    }

    const keys = this.data[0];

    return this.data.slice(1).map(row => {
      const obj = {};

      if (row.length !== keys.length) {
        throw new Error('rows must be the same length');
      }
      
      row.forEach((cell, i) => {
        obj[keys[i]] = cell;
      });

      return obj;
    });
  }

  asPairs<T = any>(mapValue?: (value: string, key?: string) => T): {[key: string]: T} {
    const obj = {};

    this.data.forEach((row) => {
      const [key, value] = row;

      if (row.length !== 2) {
        throw new Error('expected every row to have length 2');
      }
      obj[key] = mapValue ? mapValue(value, key) : value;
    });

    return obj;
  }
}
