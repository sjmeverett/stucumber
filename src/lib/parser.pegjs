{
  function expandTemplateString(template, example) {
    return template.replace(/<([^>]+)>/g, (_, key) => example[key]);
  }

  function zip(keys, values) {
    const obj = {};

    for (let i = 0; i < keys.length; i++) {
      obj[keys[i]] = values[i];
    }

    return obj;
  }

  function flatten(arr) {
    const result = [];

    for (let element of arr) {
      Array.prototype.push.apply(result, Array.isArray(element) ? element : [element]);
    }

    return result;
  }
}

Feature
  = _ annotations:Annotations TFeature name:String NL Preamble? _ scenarios:Scenarios
	{ return { name, scenarios, annotations } }

Annotation 
  = TAt attribute:Keyword _
  { return attribute }

Annotations
  = Annotation*

Preamble
  = As Want Reason
  / As Want

As
  = _ TAs actor:String NL
  { return actor }

Want
  = _ TIWant want:String NL 
  { return want }

Reason
  = _ TSo reason:String NL
  { return reason }

Scenarios
  = scenarios:Scenario*
  { return flatten(scenarios) }
          
Scenario = _ annotations:Annotations TScenario name:String NL rules:Rules _
  { return { name, rules, annotations } }

  / _ annotations:Annotations TScenarioOutline name:String NL rules:Rules examples:Examples _
  { 
    return examples.map((example) => ({
      name: {value: expandTemplateString(name.value, example), location: name.location},
      rules: rules.map((template) => ({value: expandTemplateString(template.value, example), location: template.location})),
      annotations
    }));
  }

Rules
  = rules:Rule+
  { return rules }

Rule
  = _ Clause rule:String NL table:Table
  { return Object.assign({}, rule, {table}) }

Clause
  = TGiven
  / TWhen
  / TThen
  / TAnd
  / TStar
	
Examples
  = _ TExamples NL table:Table
  {
    const keys = table[0];
    const data = [];
    const rows = table.slice(1);

    for (const row of rows) {
      const rowData = zip(keys, row);
      data.push(rowData);
    }

    return data;
  }

Table
  = rows:TableRow*
  { return rows }

TableRow
  = _ TTableSep cells:TableCell* NL
  { return cells }

TableCell
  = cell:TTableCell TTableSep
  { return cell }

TFeature = "Feature:"
TAs = "As"
TIWant = "I want"

TSo
  = "So"
	/ "In order"

TScenario = "Scenario:"
TScenarioOutline = "Scenario Outline:"
TGiven = "Given"
TWhen = "When"
TThen = "Then"
TAnd = "And"
TExamples = "Examples:"
TTableSep = "|"
TStar = "*"
TAt = "@"

TTableCell
  = data:[^|\n]+
  { return data.join('').trim() }

String
  = str:[^\n]+
  { 
    return {
      value: str.join('').trim(),
      location: location().start
    }
  }

Keyword
  = str:[a-zA-Z0-9_-]+
  { return str.join('').trim() }

NL = "\n"

_ = WS Comment _
  / WS

Comment
  = "//" [^\n]* NL
  / "#" [^\n]* NL

WS "whitespace"
  = [ \t\n\r]*
  