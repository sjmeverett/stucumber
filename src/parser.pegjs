{
  function expandTemplateString(template, example) {
    return template.replace(/<([^>]+)>/g, (_, key) => example[key]);
  }

  function zip(keys, values) {
    const obj = {};

    for (let i = 0; i < keys.length; i++) {
      obj[keys[i]] = values[i];
    }

    return obj;
  }

  function flatten(arr) {
    const result = [];

    for (let element of arr) {
      Array.prototype.push.apply(result, Array.isArray(element) ? element : [element]);
    }

    return result;
  }
}

Feature
  = _ annotations:Annotations TFeature name:String NL Preamble? ruleDeclarations:RuleDeclarations background:Background? _ scenarios:Scenarios
	{ return { name, ruleDeclarations, background, scenarios, annotations } }

Annotation
  = TAt attribute:Keyword "(" args:[^)]* ")" _
  { return { name: attribute, arguments: JSON.parse('[' + args.join('') + ']') } }
  / TAt attribute:Keyword _
  { return { name: attribute, arguments: [] } }

Annotations
  = Annotation*

Preamble
  = As Want Reason
  / As Want

As
  = _ TAs actor:String NL
  { return actor }

Want
  = _ TIWant want:String NL
  { return want }

Reason
  = _ TSo reason:String NL
  { return reason }

Background
  = _ TBackground NL rules:Rules
  { return rules }

RuleDeclarations
  = ruleDeclarations:RuleDeclaration*
  { return ruleDeclarations }

RuleDeclaration
  =  _ TRule template:String NL rules:Rules
  { return {template, rules} }

Scenarios
  = scenarios:Scenario*
  { return flatten(scenarios) }

Scenario
  = _ annotations:Annotations TScenario name:String NL rules:Rules _
  { return { name, rules, annotations } }

  / _ annotations:Annotations TScenarioOutline name:String NL rules:Rules examples:Examples _
  {
    return examples.map((example) => ({
      name: {value: expandTemplateString(name.value, example), location: name.location},
      rules: rules.map((template) => ({value: expandTemplateString(template.value, example), location: template.location, keyword: template.keyword})),
      annotations
    }));
  }

Rules
  = rules:Rule+
  { return rules }

Rule
  = _ keyword:Clause rule:String EOS data:Table
  { return Object.assign({}, rule, {data, keyword}) }

Clause
  = TGiven
  / TWhen
  / TThen
  / TAnd
  / TStar

Examples
  = _ TExamples NL table:Table
  {
    const keys = table[0];
    const data = [];
    const rows = table.slice(1);

    for (const row of rows) {
      const rowData = zip(keys, row);
      data.push(rowData);
    }

    return data;
  }

Table
  = rows:TableRow*
  { return rows }

TableRow
  = _ TTableSep cells:TableCell* EOS
  { return cells }

TableCell
  = cell:TTableCell TTableSep
  { return cell }

TFeature = "Feature:"
TAs = "As"
TIWant = "I want"

TSo
  = "So"
	/ "In order"

TScenario = "Scenario:"
TScenarioOutline = "Scenario Outline:"
TBackground = "Background:"
TRule = "Rule:"
TGiven = "Given"
TWhen = "When"
TThen = "Then"
TAnd = "And"
TExamples = "Examples:"
TTableSep = "|"
TStar = "*"
TAt = "@"

TTableCell
  = data:[^|\n]+
  {
    const cell = data.join('').trim();

    return /^"[^"]*"$/.test(cell)
      ? JSON.parse(cell)
      : cell;
  }

String
  = str:[^\n]+
  {
    return {
      value: str.join('').trim(),
      location: location().start
    }
  }

EOS
  = NL
  / EOF

EOF
  = !.

Keyword
  = str:[a-zA-Z0-9_-]+
  { return str.join('').trim() }

NL = [\n\r]

_ = WS Comment _
  / WS

Comment
  = "//" [^\n]* NL
  / "#" [^\n]* NL

WS "whitespace"
  = [ \t\n\r]*
