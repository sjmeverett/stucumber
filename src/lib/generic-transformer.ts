import Transformer from './transformer';
import {Feature, Scenario, Clause, Rule} from './parser';

// TODO: upgrade when new version is published, source-map types are currently fucked
const SourceNode = require('source-map').SourceNode;

export interface GenericTransformerOptions {
  featureFn?: string;
  scenarioFn: string;
  beforeAllFn: string;
  afterAllFn: string;
  preamble?: string;
  getFeatureName?: (feature: Feature) => string;
  getScenarioName?: (feature: Feature, scenario: Scenario) => string;
}

export default class GenericTransformer extends Transformer<any> {
  protected options: GenericTransformerOptions;

  constructor(options: GenericTransformerOptions) {
    super();

    this.options = {
      getFeatureName: (feature: Feature) => 'Feature: ' + feature.name.value,
      getScenarioName: (feature: Feature, scenario: Scenario) => scenario.name.value,
      preamble: 'const {cucumber} = require("stucumber");',
      ...options
    };
  }

  protected transformFile(filename: string, file) {
    const {code, map} = new SourceNode(1, 1, filename, [
      this.options.preamble,
      file
    ]).toStringWithSourceMap({file: filename});

    return (
      code +
      '\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,' +
      Buffer.from(map.toString(), 'utf8').toString('base64')
    );
  }

  protected transformFeature(filename: string, feature: Feature, scenarios) {
    let chunks = [
      `${this.options.beforeAllFn}(() => cucumber.enterFeature(${JSON.stringify(
        feature.annotations
      )}));`,
      `${this.options.afterAllFn}(() => cucumber.exitFeature(${JSON.stringify(
        feature.annotations
      )}));`,
      ...scenarios
    ];

    if (this.options.featureFn) {
      chunks = [
        this.applyAttributes(this.options.featureFn, feature.annotations),
        `(`,
        JSON.stringify(this.options.getFeatureName(feature)),
        `, () => {`,
        ...chunks,
        `});`
      ];
    }

    return new SourceNode(
      feature.name.location.line,
      feature.name.location.column,
      filename,
      chunks
    );
  }

  protected transformScenario(filename: string, feature: Feature, scenario: Scenario, rules) {
    return new SourceNode(scenario.name.location.line, scenario.name.location.column, filename, [
      this.applyAttributes(this.options.scenarioFn, scenario.annotations),
      `(`,
      JSON.stringify(this.options.getScenarioName(feature, scenario)),
      `, () => {`,
      `const world = cucumber.createWorld();`,
      `return cucumber.enterScenario(world, `,
      JSON.stringify([...feature.annotations, ...scenario.annotations]),
      `)`,
      ...[].concat(...rules),
      `.then(() => cucumber.exitScenario(world, `,
      JSON.stringify([...feature.annotations, ...scenario.annotations]),
      `));`,
      `});`
    ]);
  }

  protected transformRule(filename: string, feature: Feature, scenario: Scenario, rule: Rule) {
    return [
      `.then(() => `,
      new SourceNode(rule.location.line, rule.location.column, filename, [
        `cucumber.rule(world, `,
        JSON.stringify(rule.value),
        ', ',
        rule.data ? JSON.stringify(rule.data) : 'null',
        `)`
      ]),
      `)`
    ];
  }

  protected applyAttributes(name: string, attributes: string[]) {
    let attribute = '';

    if (attributes.indexOf('skip') > -1) {
      attribute = '.skip';
    } else if (attributes.indexOf('only') > -1) {
      attribute = '.only';
    }

    return name + attribute;
  }
}
