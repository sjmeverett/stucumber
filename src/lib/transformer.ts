import {Feature, Scenario, Clause, parse, Rule} from './parser';

export default abstract class Transformer<T> {
  protected abstract transformFeature(filename: string, feature: Feature, scenarios: T[]): T;

  protected abstract transformScenario(
    filename: string,
    feature: Feature,
    scenario: Scenario,
    rules: T[]
  ): T;

  protected abstract transformRule(
    filename: string,
    feature: Feature,
    scenario: Scenario,
    rule: Rule
  ): T;
  
  protected abstract transformFile(filename: string, file: T): string;

  transform(filename: string, source: string);
  transform(filename: string, feature: Feature);
  transform(filename: string, source: string | Feature) {
    const feature = typeof source === 'string' ? parse(source) : source;

    return this.transformFile(
      filename,
      this.transformFeature(
        filename,
        feature,
        feature.scenarios.map((scenario) =>
          this.transformScenario(
            filename,
            feature,
            scenario,
            scenario.rules.map((rule) => this.transformRule(filename, feature, scenario, rule))
          )
        )
      )
    );
  }
}
