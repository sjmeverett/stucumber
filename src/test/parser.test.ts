import { parse } from "../lib/parser";
import { readFileSync } from "fs";

describe("Parser", () => {
  it("should parse a standard scenario", () => {
    const result = parse(readFileSync(__dirname + '/scenario.feature', 'utf8'));
    expect(result).toMatchSnapshot();
  });

  it("should parse a scenario outline", () => {
    const result = parse(readFileSync(__dirname + '/scenario-outline.feature', 'utf8'));
    expect(result).toMatchSnapshot();
  });

  it("should parse a scenario with annotations", () => {
    const result = parse(readFileSync(__dirname + '/annotations.feature', 'utf8'));
    expect(result).toMatchSnapshot();
  });

  it("should parse a scenario with data tables", () => {
    const result = parse(readFileSync(__dirname + '/data-table.feature', 'utf8'));
    expect(result).toMatchSnapshot();
  });  
});
