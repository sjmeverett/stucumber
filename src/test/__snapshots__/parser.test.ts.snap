// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser should parse a scenario outline 1`] = `
Object {
  "annotations": Array [],
  "name": Object {
    "location": Object {
      "column": 9,
      "line": 1,
      "offset": 8,
    },
    "value": "foo",
  },
  "scenarios": Array [
    Object {
      "annotations": Array [],
      "name": Object {
        "location": Object {
          "column": 20,
          "line": 2,
          "offset": 32,
        },
        "value": "this is a scenario outline",
      },
      "rules": Array [
        Object {
          "location": Object {
            "column": 10,
            "line": 3,
            "offset": 69,
          },
          "value": "that I have abc",
        },
        Object {
          "location": Object {
            "column": 9,
            "line": 4,
            "offset": 98,
          },
          "value": "I 123",
        },
        Object {
          "location": Object {
            "column": 9,
            "line": 5,
            "offset": 116,
          },
          "value": "I -Â£*%",
        },
      ],
    },
  ],
}
`;

exports[`Parser should parse a scenario with annotations 1`] = `
Object {
  "annotations": Array [
    "annotation1",
  ],
  "name": Object {
    "location": Object {
      "column": 9,
      "line": 2,
      "offset": 21,
    },
    "value": "foo",
  },
  "scenarios": Array [
    Object {
      "annotations": Array [
        "annotation2",
        "annotation3",
      ],
      "name": Object {
        "location": Object {
          "column": 12,
          "line": 5,
          "offset": 67,
        },
        "value": "this is a test",
      },
      "rules": Array [
        Object {
          "location": Object {
            "column": 6,
            "line": 6,
            "offset": 88,
          },
          "table": Array [],
          "value": "a thing",
        },
      ],
    },
  ],
}
`;

exports[`Parser should parse a scenario with data tables 1`] = `
Object {
  "annotations": Array [],
  "name": Object {
    "location": Object {
      "column": 9,
      "line": 1,
      "offset": 8,
    },
    "value": "data tables",
  },
  "scenarios": Array [
    Object {
      "annotations": Array [],
      "name": Object {
        "location": Object {
          "column": 12,
          "line": 2,
          "offset": 32,
        },
        "value": "data table",
      },
      "rules": Array [
        Object {
          "location": Object {
            "column": 10,
            "line": 3,
            "offset": 53,
          },
          "table": Array [
            Array [
              "datum1",
              "datum2",
            ],
            Array [
              "datum3",
              "datum4",
            ],
          ],
          "value": "I have data",
        },
      ],
    },
  ],
}
`;

exports[`Parser should parse a standard scenario 1`] = `
Object {
  "annotations": Array [],
  "name": Object {
    "location": Object {
      "column": 9,
      "line": 2,
      "offset": 30,
    },
    "value": "foo",
  },
  "scenarios": Array [
    Object {
      "annotations": Array [],
      "name": Object {
        "location": Object {
          "column": 12,
          "line": 8,
          "offset": 117,
        },
        "value": "this is a test",
      },
      "rules": Array [
        Object {
          "location": Object {
            "column": 10,
            "line": 9,
            "offset": 142,
          },
          "table": Array [],
          "value": "that I am testing",
        },
        Object {
          "location": Object {
            "column": 8,
            "line": 10,
            "offset": 168,
          },
          "table": Array [],
          "value": "I am still testing",
        },
        Object {
          "location": Object {
            "column": 9,
            "line": 11,
            "offset": 196,
          },
          "table": Array [],
          "value": "I test",
        },
        Object {
          "location": Object {
            "column": 8,
            "line": 12,
            "offset": 211,
          },
          "table": Array [],
          "value": "continue testing",
        },
        Object {
          "location": Object {
            "column": 9,
            "line": 13,
            "offset": 237,
          },
          "table": Array [],
          "value": "I want to test",
        },
        Object {
          "location": Object {
            "column": 8,
            "line": 14,
            "offset": 260,
          },
          "table": Array [],
          "value": "test",
        },
      ],
    },
  ],
}
`;
